# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmTypes.cpp
# Commit: a850ff24c2ae
# Full Hash: a850ff24c2aeaefac447d9e501f6e443f395f844
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2020-08-18 09:24:52
# Regressor Bug: 1561521
# File Overlap Count: 1
# Description:
#   Bug 1561521 - Clean up wasm::ToString and drop wasm::ToCString. r=lth
#   
#   wasm::ToCString is losing relevance as we have more constructed types. This
#   commit removes ToCString and replaces all uses with wasm::ToString.
#   
# ==============================================================================

diff -r 3d1a8283d2f8 -r a850ff24c2ae js/src/wasm/WasmTypes.cpp
--- a/js/src/wasm/WasmTypes.cpp	Tue Aug 18 00:07:29 2020 +0000
+++ b/js/src/wasm/WasmTypes.cpp	Tue Aug 18 00:45:24 2020 +0000
@@ -1016,17 +1016,38 @@
       literal = "f64";
       break;
     case ValType::Ref:
-      switch (type.refTypeKind()) {
-        case RefType::Extern:
-          literal = "externref";
-          break;
-        case RefType::Func:
-          literal = "funcref";
-          break;
-        case RefType::TypeIndex:
-          return JS_smprintf("optref %d", type.refType().typeIndex());
+      if (type.isNullable() && !type.isTypeIndex()) {
+        switch (type.refTypeKind()) {
+          case RefType::Extern:
+            literal = "externref";
+            break;
+          case RefType::Func:
+            literal = "funcref";
+            break;
+          case RefType::TypeIndex:
+            MOZ_ASSERT_UNREACHABLE();
+        }
+      } else {
+        const char* heapType = nullptr;
+        switch (type.refTypeKind()) {
+          case RefType::Extern:
+            heapType = "externref";
+            break;
+          case RefType::Func:
+            heapType = "funcref";
+            break;
+          case RefType::TypeIndex:
+            return JS_smprintf("ref %s%d", type.isNullable() ? "null " : " ",
+                               type.refType().typeIndex());
+        }
+        return JS_smprintf("ref %s%s", type.isNullable() ? "null " : " ",
+                           heapType);
       }
       break;
   }
   return JS_smprintf("%s", literal);
 }
+
+UniqueChars wasm::ToString(const Maybe<ValType>& type) {
+  return type ? ToString(type.ref()) : JS_smprintf("%s", "void");
+}